Add support for OpenSSL 1.1.0

alexcrichton opened the pull request:
This commit is relatively major refactoring of the `openssl-sys` crate as well
as the `openssl` crate itself. The end goal here was to support OpenSSL 1.1.0,
and lots of other various tweaks happened along the way. The major new features
are:

* OpenSSL 1.1.0 is supported
* OpenSSL 0.9.8 is no longer supported (aka all OSX users by default)
* All FFI bindings are verified with the `ctest` crate (same way as the `libc`
  crate)
* CI matrixes are vastly expanded to include 32/64 of all platforms, more
  OpenSSL version coverage, as well as ARM coverage on Linux
* The `c_helpers` module is completely removed along with the `gcc` dependency.
* The `openssl-sys` build script was completely rewritten
  * Now uses `OPENSSL_DIR` to find the installation, not include/lib env vars.
  * Better error messages for mismatched versions.
  * Better error messages for failing to find OpenSSL on a platform (more can be
    done here)
  * Probing of OpenSSL build-time configuration to inform the API of the `*-sys`
    crate.
* Many Cargo features have been removed as they're now enabled by default.

As this is a breaking change to both the `openssl` and `openssl-sys` crates this
will necessitate a major version bump of both. There's still a few more API
questions remaining but let's hash that out on a PR!

Closes #452

alexcrichton commented on openssl-sys/build.rs:
```diff
@@ -1,86 +1,323 @@
+    // TODO: try to be more helpful in this message, e.g.:
+    //
+    // * if we're on linux, recommend apt-get something or other
+    // * if we're on Windows, link to a README explaining more
```
I'd love to handle this specifically (compiling on Windows) before landing. Do you perhaps have suggestions about what to put in this error though?

sfackler replied:
I actually have never had success getting a MinGW-installed OpenSSL to work - the compiler seems to have troubles finding linux-style libraries or something. The README currently points people to http://slproweb.com/products/Win32OpenSSL.html to obtain binaries.

alexcrichton replied:
Hm the MinGW packages seem to be working on AppVeyor at least? I suppose I can otherwise just point to the README on Windows.

sfackler replied:
Oh great, if that works it seems preferable.

alexcrichton commented on openssl-sys/src/ossl110.rs:
```diff
@@ -0,0 +1,142 @@
+pub const CRYPTO_EX_INDEX_SSL_CTX: c_int = 1;
+
+pub fn init() {
+    // TODO: fill this in, probably need to at least do something?
```
Oh I never did get around to filling this out. Do you know if OpenSSL 1.1.0 actually needs any initialization?

jwilm replied:
[`OPENSSL_init_ssl()`](https://www.openssl.org/docs/man1.1.0/ssl/OPENSSL_init_ssl.html) and [`OPENSSL_init_crypto()`](https://www.openssl.org/docs/man1.1.0/crypto/OPENSSL_init_crypto.html) both say

> As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required. Similarly it will also automatically deinitialise as required.

It seems like as long as the defaults are ok, then no explicit call is necessary. Those man pages list the defaults.

jwilm replied:
Just kidding... now that I've looked more closely, this is actually just for 1.1.0.

alexcrichton replied:
Oh yay! Sounds like we don't actually have to do anything in that case, thanks @jwilm!

alexcrichton commented on openssl/src/ssl/bio.rs:
```diff
@@ -78,24 +66,16 @@ pub unsafe fn get_mut<'a, S: 'a>(bio: *mut BIO) -> &'a mut S {
     ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(f))
 }
 
-#[cfg(not(feature = "nightly"))]
```
I took the liberty to remove this here

alexcrichton commented on openssl/src/ssl/bio.rs:
```diff
@@ -175,22 +155,128 @@ unsafe extern "C" fn ctrl<S: Write>(bio: *mut BIO,
+    pub use ffi::{BIO_set_init, BIO_set_flags, BIO_set_data, BIO_get_data};
+
+    pub unsafe fn BIO_set_num(_bio: *mut ffi::BIO, _num: c_int) {
+        // TODO: what should this do?
```
I don't even know what the `num` field was previously, @sfackler do you know if it's important to fill this in?

sfackler replied:
I believe it's just a utility field that the BIO implementation can use to store stuff. If we're not using it it should be fine to ignore.

alexcrichton commented on openssl/src/ssl/mod.rs:
```diff
@@ -73,74 +72,11 @@ bitflags! {
+    /// Support the TLS protocol
+    Tls,
+    /// Support DTLS protocol
+    Dtls,
```
The size of this enum both before and after worries me. Do you have thoughts though of how to proceed here?

sfackler replied:
Are the other methods no longer exposed in 1.1?

alexcrichton replied:
All of the `FOOvX` methods were deprecated, but still exist. I believe the SSLv2 method was outright deleted.

sfackler replied:
Cool, that makes sense. We should make sure to add the other `SSL_OP_NO_*` constants and have people configure protocol selection through those.

sfackler commented on openssl/src/crypto/rsa.rs:
```diff
@@ -42,14 +42,14 @@ impl Drop for RSA {
     /// only useful for associating the key material directly with the key, it's safer to use
     /// the supplied load and save methods for DER formatted keys.
-    pub fn from_public_components(n: BigNum, e: BigNum) -> Result<RSA, ErrorStack> {
+    pub fn from_public_components(n: BigNum,
```
Wait, `d` is the private exponent. I don't think it makes sense to pass it in here.

alexcrichton replied:
Oh I have no idea what it is, we just needed to pass something to the API on 1.1.0. I guess that should be null?

sfackler replied:
That's what I'd guess?

sfackler commented on openssl/src/ssl/mod.rs:
```diff
@@ -1412,3 +1343,107 @@ impl<'a> IntoSsl for &'a SslContext {
+                                     Some(f))
+    }
+
+    pub unsafe fn get_method(method: SslMethod) -> *const ffi::SSL_METHOD {
```
Is this used?

alexcrichton replied:
In one location yeah to do this conversion when calling `SSL_CTX_new`

sfackler replied:
Oh right

sfackler commented on openssl/src/ssl/tests/mod.rs:
```diff
@@ -50,18 +51,28 @@ struct Server {
+        let cert = me.join("cert.pem");
+        let key = me.join("key.pem");
+
+        File::create(&cert).unwrap().write_all(CERT).unwrap();
```
Could we stick these in a tmpdir?

alexcrichton replied:
You ok picking that up as a dep? I figured this was as good a place as any without picking up new dependencies.

sfackler replied:
It's just a dev dependency. w/e

sfackler commented on openssl/src/ssl/tests/mod.rs:
```diff
@@ -226,10 +237,10 @@ run_test!(new_sslstream, |method, stream| {
-    let ssl = Ssl::new(&SslContext::new(method).unwrap()).unwrap();
-    assert_eq!(ssl.ssl_method(), method);
-});
+// run_test!(get_ssl_method, |method, _| {
```
Dead code?

alexcrichton replied:
Oh I left this in as I figure we'd want to sort out what to do with the `ssl_method` method, which this was testing.

sfackler replied:
Ah right. Let's leave ssl_method out entirely for now and see if anyone asks for it back. The implementation always seemed a bit sketchy to me with the unwrap.

sfackler commented on systest/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+openssl-sys = { path = "../openssl-sys" }
+
+[build-dependencies]
+ctest = { git = "https://github.com/alexcrichton/ctest" }
```
It'd be kind of nice if this were published.

sfackler commented on systest/build.rs:
```diff
@@ -0,0 +1,111 @@
+    });
+    cfg.skip_fn(move |s| {
+        s == "CRYPTO_memcmp" ||                 // uses volatile
+            s == "DH_new_from_params" ||        // not defined
```
This should just be removed. I missed it by accident a while ago.

sfackler commented on systest/build.rs:
```diff
@@ -0,0 +1,111 @@
+        (s == "EVP_PKEY" && field == "pkey") ||      // union
+            (s == "GENERAL_NAME" && field == "d")    // union
+    });
+    cfg.skip_signededness(|s| {
```
Why are we skipping signedness checks?

alexcrichton replied:
Oh this is either a bug or a feature in ctest depending on how you look at it. It tries to test the signedness of all `type foo = bar` items in Rust, but these types skipped are all function pointers, which have no signedness.

sfackler commented:
Worth noting that this also drops 1.0.0 support, but that also seems fine since it's no longer supported either.

jwilm commented:
@sfackler I haven't reviewed this PR in detail, but at the surface, it appears to support 1.0.x given that files like [openssl-sys/src/ossl10x.rs](https://github.com/sfackler/rust-openssl/pull/464/files#diff-e54f7c19f943634f9bee0771e488e7c1) are included. How do you figure otherwise?

sfackler commented:
I'm talking about 1.0.0 specifically, not 1.0.1 or 1.0.2. See here for the check: https://github.com/sfackler/rust-openssl/pull/464/files#diff-aca849e774a4e9d79fcc036a9de0d596R212

jwilm commented:
Ah, thanks for the clarification.

cuviper commented on openssl-sys/build.rs:
```diff
@@ -1,86 +1,323 @@
-    	"dylib"
+    // Do a bit of string parsing to find `#define OPENSSL_VERSION_TEXT "..."`
+    let version_line = version_header.lines().find(|l| {
+        l.contains("OPENSSL_VERSION_TEXT") && !l.contains("fips")
```
There are multiple lines, e.g. I have:

```
# define OPENSSL_VERSION_NUMBER  0x1010002fL
# ifdef OPENSSL_FIPS
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0b-fips  26 Sep 2016"
# else
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0b  26 Sep 2016"
# endif
```

But it might be easier to work with `OPENSSL_VERSION_NUMBER`.

sfackler replied:
Ah that makes sense. Yeah, working with VERSION_NUMBER does sound a bit more reliable.

alexcrichton replied:
The "wrong version found" message is now a bit worse, but the version detection is likely more robust.

alexcrichton commented on openssl-sys/Cargo.toml:
```diff
@@ -10,38 +10,12 @@ documentation = "https://sfackler.github.io/rust-openssl/doc/v0.7.17/openssl_sys
 [build-dependencies]
 pkg-config = "0.3"
 
-[target.le32-unknown-nacl.dependencies]
```
For now, yeah, I figured this was just from a time long since passed.

alexcrichton commented on openssl-sys/build.rs:
```diff
@@ -1,86 +1,323 @@
+        }
+
+        let system = Path::new("/usr/lib/libssl.0.9.8.dylib");
+        if system.exists() {
```
Hm good point, we should probably do that. `pkg-config` is somewhat nonstandard on OSX I think, but no harm in running it.

sfackler replied:
The picture I have in my head is that the OSX codepath behaves just like the Linux codepath except that it first checks for a Homebrew install.

alexcrichton replied:
👍

alexcrichton commented on openssl-sys/build.rs:
```diff
@@ -1,86 +1,323 @@
+    if lib.include_paths.len() == 0 {
+        panic!("
+
+Used pkg-config to discover the OpenSSL installation, but pkg-config did not
```
In theory `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS` set above will force `pkg-config` to print all those out, but if that doesn't work then yeah this'll be a weird failure.

alexcrichton commented on openssl-sys/src/lib.rs:
```diff
@@ -374,30 +154,21 @@ pub const SSL_VERIFY_NONE: c_int = 0;
-pub const SSL_OP_NO_SSLv2: c_long =                                 0x01000000;
-pub const SSL_OP_NO_SSLv3: c_long =                                 0x02000000;
-pub const SSL_OP_NO_TLSv1: c_long =                                 0x04000000;
+#[cfg(not(ossl101))]
```
I was thinking that for a particular OpenSSL version this gives you what *can* be the API of that version, although it'd ideally be trimmed to the precise API of what's exactly on the system.

In that sense I'm definitely in favor of conditionally compiling away all the OpenSSL 1.1.0 APIs when compiling for OpenSSL 1.0.1, for example, to weed out errors sooner.

(also this is required to pass systest)

alexcrichton commented on openssl-sys/src/lib.rs:
```diff
@@ -1098,14 +719,30 @@ extern "C" {
-    pub fn sk_free(st: *mut _STACK);
-    pub fn sk_pop_free(st: *mut _STACK, free: Option<unsafe extern "C" fn (*mut c_void)>);
-    pub fn sk_pop(st: *mut _STACK) -> *mut c_char;
+    #[cfg_attr(any(ossl101, ossl102), link_name = "sk_free")]
```
How so? It's a bit of a weird attribute but allows us to work with OpenSSL 1.1.0 idioms while still working against older versions.

sfackler replied:
It doesn't look like the usage of these function's can't be unified due to the different representation of `_STACK`, so it doesn't seem like munging the name of the function on 1.0.x buys us anything other than confusing people searching for the 1.0.x name.

alexcrichton replied:
Ok, I can conditionally compile it

alexcrichton commented on openssl-sys/src/ossl10x.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+#[repr(C)]
+pub struct X509 {
```
These were needed for defining various functions here and there, and they're all validated as being correct. It may be possible to configure OpenSSL in a number of configurations perhaps to test, but the definitions here were all at least copied from the actual header file.

alexcrichton commented on openssl-sys/src/ossl110.rs:
```diff
@@ -0,0 +1,142 @@
+pub enum X509 {}
+pub enum SSL_CTX {}
+
+pub const SSL_OP_MICROSOFT_SESS_ID_BUG: c_ulong =                   0x00000000;
```
Apparently, yes!

alexcrichton commented on openssl/src/bn/mod.rs:
```diff
@@ -185,53 +185,54 @@ impl<'a> BigNumRef<'a> {
     }
 
-    pub fn sub_word(&mut self, w: c_ulong) -> Result<(), ErrorStack> {
+    pub fn sub_word(&mut self, w: u32) -> Result<(), ErrorStack> {
```
This was actually an incorrect definition of the `BN_ULONG` type alias. (evidenced by the changes in the sys crate above). The definition of this is `c_uint` on Windows which is only 32-bits.

I figured it's safe at least to consume the *smallest* bit width used across platforms and then we'd return the largest bit width any platform returns. In that sense there shouldn't be a downcast?

alexcrichton commented on openssl/src/crypto/hash.rs:
```diff
@@ -123,22 +121,20 @@ impl Hasher {
+        unsafe {
+            try_ssl!(ffi::EVP_DigestUpdate(self.ctx,
+                                           data.as_ptr() as *mut _,
+                                           data.len()));
```
I believe so, yeah. The systest crate turned this up at least and the actual definition uses `size_t` on all versions.

alexcrichton commented on openssl/src/crypto/hmac.rs:
````diff
@@ -66,7 +64,7 @@ use self::State::*;
 /// assert_eq!(res, spec);
 /// ```
 pub struct HMAC {
````
Due to the size of this, perhaps that could be done in a follow-up PR?

sfackler replied:
Yeah, seems fine.

alexcrichton commented on openssl/src/crypto/hmac.rs:
```diff
@@ -202,6 +190,73 @@ pub fn hmac(t: Type, key: &[u8], data: &[u8]) -> Result<Vec<u8>, ErrorStack> {
+
+    use ffi;
+
+    pub struct HMAC_CTX {
```
I figured it's good hygiene to return `*mut T` from a `&self`, but not required other than that.

sfackler replied:
Oh right, the struct is stored inline rather than being allocated by OpenSSL.

All of this stuff'll be deleted when we switch over to the PKey impl so it doesn't really matter.

sfackler commented on openssl-sys/build.rs:
```diff
@@ -64,46 +64,69 @@ fn find_openssl_dir(target: &str) -> OsString {
 
 As a consumer of this crate, you can fix this error by using Homebrew to
 install the `openssl` package, or as a maintainer you can use the openssl-sys
 0.7 crate for support with OpenSSL 0.9.8.
```
You'd actually want to use openssl-sys 0.8, since 0.7 required headers.

sfackler replied:
Oh right derp, was thing of openssl not openssl-sys.

alexcrichton commented:
Ok, lots of commits addressing comments pushed. I'm curious though, could you elaborate on this point?

> Cool, that makes sense. We should make sure to add the other SSL_OP_NO_* constants and have people configure protocol selection through those.

sfackler commented:
I'm talking about these constants: https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/src/lib.rs#L398-L400.

OpenSSL also defines ones to turn of TLSv1.1 and TLSv1.2, and some DTLS variants. We'd want to add those as well so people can e.g. only support TLSv1.2

sfackler commented:
Oh right, they're sitting just below commented out: https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/src/lib.rs#L402

sfackler commented:
(The'll also need to then be added to the bitflags thing in the ssl module).

sfackler commented:
Seems like some cfgs are missing based off of travis.

alexcrichton commented:
Ok adding cfgs to try to fix tests, updated README, addressed more comments, still missing SSL_OP business.

If you know what to do there it may be best if you just send a commit for that?

alexcrichton commented:
@sfackler I had to add [this commit](https://github.com/sfackler/rust-openssl/commit/989b28012a912b8175e9a142320a087e4b5697c4) to prevent segfaults in the `test_alpn_server_select_none` test which is failing like:

```
failures:

---- ssl::tests::test_alpn_server_select_none stdout ----

	thread 'ssl::tests::test_alpn_server_select_none' panicked at 'bad reason: 336151648', openssl/src/error.rs:142

failures:

    ssl::tests::test_alpn_server_select_none
```

Any ideas?

sfackler commented:
Not sure what'd be going on there. It could be that error handling is different in 1.1

sfackler commented:
Opened a PR to add the remaining constants.

sfackler commented on openssl-sys/src/lib.rs:
```diff
@@ -264,35 +73,10 @@ pub struct X509V3_CTX {
-impl Clone for BIGNUM {
-    fn clone(&self) -> BIGNUM { *self }
-}
+#[cfg(target_pointer_width = "64")]
```
The conditions in openssl's definition seem different/more complex. Are these equivalent? https://github.com/openssl/openssl/blob/6f0ac0e2f27d9240516edb9a23b7863e7ad02898/crypto/bn/asm/x86_64-gcc.c#L67

alexcrichton replied:
Yeah this wasn't 100% the same from what I could tell but I figured it was close enough in terms of ctest should turn up any problems.

sfackler commented on openssl/Cargo.toml:
```diff
@@ -12,40 +12,33 @@ build = "build.rs"
-pkcs5_pbkdf2_hmac = ["openssl-sys/pkcs5_pbkdf2_hmac"]
-hmac_clone = ["openssl-sys/hmac_clone"]
+aes_xts = []
+aes_ctr = []
```
These were added in 1.0.2 as well?

alexcrichton commented:
Ok I pushed one last commit to just ignore the failing test on OpenSSL 1.1.0, I may not have enough time to track it down right now :(

I think that should be everything though?

sfackler commented on openssl/Cargo.toml:
```diff
@@ -12,40 +12,33 @@ build = "build.rs"
+# Added in OpenSSL 1.0.2
+rfc5114 = []
+
+# TODO: what to do about these features?
```
We can delete these I believe.

alexcrichton commented on .travis.yml:
```diff
@@ -1,28 +1,86 @@
+cache:
+  cargo: true
+  directories:
+    - $HOME/openssl
```
Oh this was actually just caching the OpenSSL we compiled. Figured it'd make the builds a bit faster and less noisy to avoid always doing that.

sfackler replied:
Oh right.

sfackler commented:
The ALPN failures are somewhat concerning but it seems okay to land this and figure those out later.

sfackler commented:
I've started a project to track other work that needs to happen before we can release: https://github.com/sfackler/rust-openssl/projects/1

sfackler commented:
I've added you as a collaborator on the repo so feel free to add stuff to that as needed.

sfackler commented:
Everything's working now except for DTLS tests, which are failing on Windows and ARM. Not sure what's going on there.

sfackler commented:
DTLS tests have never been run on Windows, so I'm not particularly worried about them. Not sure what's going on with the ARM failures though.

alexcrichton commented:
Looks like ARM failures are all related to [this timeout](https://github.com/sfackler/rust-openssl/blob/master/openssl/src/ssl/tests/mod.rs#L121). I increased that to 1s and the tests passed.

Perhaps we could test the client and server code at the same time? Is there a critical reason we *must* test against `openssl` itself?

sfackler commented:
Ah, the timeout thing makes sense!

I don't think there's any intrinsic reason to use `openssl` - it's probably that way because the tests were written before the server side logic existed.

sfackler commented:
Woo, merging!
